[Driver] Darwin: Link in the profile runtime archive first

vedantk opened the pull request:
While building a project with code coverage enabled, we can link in
dependencies which export a weak definition of __llvm_profile_filename.

After r306710, linking in the profiling runtime could pull in a weak
definition of this symbol from a dependency, instead of from within the
runtime's archive.

This inconsistency causes issues during API verification, and is also a
practical problem (the symbol would go missing were the dependent dylib
to be switched out). Introduce a LinkFirst runtime link option to make
sure we always search the profiling runtime for this symbol first.

See: https://reviews.llvm.org/D35385

rdar://problem/33271080

vedantk commented:
@swift-ci Please smoke test OSX platform

jrose-apple reviewed (commented):
Please do not do this. Just move the profile library checking code up.

jrose-apple commented:
Also, no reason to just do this on master-next. It's better to keep them in sync.

vedantk commented:
> Please do not do this. Just move the profile library checking code up.

I'm not sure why that would be better. I think the current approach is less fragile in the face of random code movement.

> Also, no reason to just do this on master-next. It's better to keep them in sync.

The associated clang change isn't meant for the stable branch, so putting the change on master-next should keep the behavior of swift and clang in sync.

jrose-apple commented:
> I'm not sure why that would be better. I think the current approach is less fragile in the face of random code movement.

That's exactly why I don't want to do it. Right now you can see the order of arguments by reading the code. If parts of it start inserting themselves before or after other parts, it'll become a complete mess instead of just a partial mess.

> The associated clang change isn't meant for the stable branch, so putting the change on master-next should keep the behavior of swift and clang in sync.

If the change doesn't *harm* anything, we'd still prefer to take it on master.

vedantk committed:
build-script: Pass UBSan options to the lldb xcodebuild invocation

Now, `./build-script --enable-ubsan` will do the right thing if you have
lldb checked out, pending an additional lldb change.

rdar://36630617
